### PR TITLE
Don't use MMU in M mode - https://github.com/riscv/riscv-openocd/issu…

### PR DESCRIPTION
**Pull request by TommyMurphyTM1234**
Fix for https://github.com/riscv/riscv-openocd/issues/474

### CODE REVIEW

File: src/target/riscv/riscv.c [hunk 1]
```diff
@@ -1383,15 +1383,26 @@ static int riscv_mmu(struct target *target, int *enabled)
 	if (riscv_rtos_enabled(target))
 		riscv_set_current_hartid(target, target->rtos->current_thread - 1);
 
-	riscv_reg_t value;
-	if (riscv_get_register(target, &value, GDB_REGNO_SATP) != ERROR_OK) {
+	/* Don't use MMU in explicit or effective M (machine) mode */
+	riscv_reg_t priv;
+	riscv_reg_t mstatus;
+	if ((riscv_get_register(target, &priv, GDB_REGNO_PRIV) == ERROR_OK) &&
+	    (riscv_get_register(target, &mstatus, GDB_REGNO_MSTATUS) == ERROR_OK) &&
```
Comment (timsifive) [inline src/target/riscv/riscv.c]: priv (part of DCSR) and mstatus are both required to be implemented, so if we can't read them I think we should return error here.

File: src/target/riscv/riscv.c [hunk 2]
```diff
@@ -1383,15 +1383,26 @@ static int riscv_mmu(struct target *target, int *enabled)
...
+	    ((get_field(mstatus, MSTATUS_MPRV) ? get_field(mstatus, MSTATUS_MPP) : priv) == PRV_M)) {
+		LOG_DEBUG("SATP/MMU ignored in M mode.");
```
Comment (timsifive) [inline src/target/riscv/riscv.c]: Can you add the value of mstatus to this error message? That way I'll at least have something to go on when somebody reports they got this error message for no good reason.

**Review by timsifive — CHANGES_REQUESTED** [review]
Thanks for addressing this, Tommy.

What testing did you do?

**Comment by timsifive** [comment]
As to your performance impact questions. There are two more register reads, but both dcsr and mstatus reads should be cached. You should be able to confirm this easily by looking at the debug log, because riscv_mmu() is called every time OpenOCD starts any memory access.

**Comment by TommyMurphyTM1234** [comment]
Thanks for the comments and feedback @timsifive.
I will address the issues that you mention above and resubmit it.
I have only tested in M mode so not when the target is in any lower privilege mode - explicitly or via mstatus.mprv/mpp.

**Review by timsifive — APPROVED** [review]
Thanks, Tommy! I'll merge this next week.

**Comment by timsifive** [comment]
I just confirmed that all riscv-tests/debug still pass.

**Comment by TommyMurphyTM1234** [comment]
Thanks a lot @timsifive.

**Pull request merged**